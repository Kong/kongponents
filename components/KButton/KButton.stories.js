import Vue from 'vue'
import { storiesOf } from '@storybook/vue'
import VueInfoAddon from 'storybook-addon-vue-info'

import KButton from './KButton.vue'

storiesOf('KButton', module)
  .addDecorator(VueInfoAddon)
  .add('Rounded False', () => ({
    components: { KButton },
    template: `<KButton :rounded="false">not round</KButton>`
  }))
  .add('Rounded True', () => ({
    components: { KButton },
    template: '<KButton :rounded="true">rounded</KButton>'
  }))
