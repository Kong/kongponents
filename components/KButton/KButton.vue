<template>
  <button class="button" :style="{color: color, borderColor: color}" @click="handleClick" :class="{rounded: rounded}">
    <slot>Hello</slot>
  </button>
</template>

<script>
export default {
  name: 'KButton',
  props: {
    rounded: Boolean,
    default: 'false',
    handleClick: {
      default: () => () => null
    },
    color: {
      default: '#42b983'
    }
  }
}
</script>

<style>
.rounded {
  border-radius: 5px;
}
.button {
  border: 3px solid;
  padding: 10px 20px;
  background-color: white;
  outline: none;
}
</style>
