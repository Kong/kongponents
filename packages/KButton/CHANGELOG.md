# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.1.16](https://github.com/Kong/kongponents/compare/@kongponents/kbutton@0.1.15...@kongponents/kbutton@0.1.16) (2020-03-06)

**Note:** Version bump only for package @kongponents/kbutton





## [0.1.15](https://github.com/Kong/kongponents/compare/@kongponents/kbutton@0.1.13...@kongponents/kbutton@0.1.15) (2020-01-22)

**Note:** Version bump only for package @kongponents/kbutton





## [0.1.12](https://github.com/Kong/kongponents/compare/@kongponents/kbutton@0.1.10-7-beta.1...@kongponents/kbutton@0.1.12) (2020-01-16)

**Note:** Version bump only for package @kongponents/kbutton





## [0.1.10-7-beta.1](https://github.com/Kong/kongponents/compare/@kongponents/kbutton@0.1.10-7...@kongponents/kbutton@0.1.10-7-beta.1) (2020-01-15)

**Note:** Version bump only for package @kongponents/kbutton





## [0.1.10-7](https://github.com/Kong/kongponents/compare/@kongponents/kbutton@0.1.10-6...@kongponents/kbutton@0.1.10-7) (2019-12-16)

**Note:** Version bump only for package @kongponents/kbutton





## [0.1.10-6](https://github.com/Kong/kongponents/compare/@kongponents/kbutton@0.1.10-5...@kongponents/kbutton@0.1.10-6) (2019-11-22)


### Bug Fixes

* **KButton:** resets focus state to default ([#173](https://github.com/Kong/kongponents/issues/173)) ([21a99ee](https://github.com/Kong/kongponents/commit/21a99eeb20f0ca6ec669329bf2e3a05e06443d2b))





## [0.1.10-5](https://github.com/Kong/kongponents/compare/@kongponents/kbutton@0.1.10-4...@kongponents/kbutton@0.1.10-5) (2019-11-12)

**Note:** Version bump only for package @kongponents/kbutton





## [0.1.10-4](https://github.com/Kong/kongponents/compare/@kongponents/kbutton@0.1.10-3...@kongponents/kbutton@0.1.10-4) (2019-11-11)

**Note:** Version bump only for package @kongponents/kbutton





## [0.1.10-3](https://github.com/Kong/kongponents/compare/@kongponents/kbutton@0.1.10-2...@kongponents/kbutton@0.1.10-3) (2019-11-08)


### Bug Fixes

* **dist:** dist to all packages ([#152](https://github.com/Kong/kongponents/issues/152)) ([217c46b](https://github.com/Kong/kongponents/commit/217c46b02afe960f8fa597ba77410dcf034b1716))





## [0.1.10-2](https://github.com/Kong/kongponents/compare/@kongponents/kbutton@0.1.10-1...@kongponents/kbutton@0.1.10-2) (2019-11-08)

**Note:** Version bump only for package @kongponents/kbutton





# @kongponents/kbutton

## Table of Contents

- [0.1.10](#0110)

### 0.1.10
 - üí•Ô∏è**Breaking Change**  Update `.button` class to a more unique prefixed `.k-button`. [#104](https://github.com/Kong/kongponents/pull/104)
