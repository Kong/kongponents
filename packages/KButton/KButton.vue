<template>
  <a
    v-if="typeof to === 'string'"
    :type="type"
    :href="to"
    :class="[size === 'default' ? '' : size, {'icon-btn': !hasText && hasIcon}, appearance]"
    class="button"
    v-on="listeners">
    <slot name="icon" /><slot/>
  </a>
  <component
    v-else
    :type="type"
    :is="buttonType"
    :to="to"
    :class="[size === 'default' ? '' : size, {'icon-btn': !hasText && hasIcon}, appearance]"
    class="button"
    v-on="listeners">
    <slot name="icon" /><slot/>
  </component>
</template>

<script>
export const appearances = {
  primary: 'primary',
  danger: 'danger',
  secondary: 'secondary',
  outlinePrimary: 'outline-primary',
  outlineDanger: 'outline-danger',
  btnLink: 'btn-link',
  btnLinkDanger: 'btn-link-danger'
}

export default {
  name: 'KButton',
  props: {
    /**
      * Base styling of the button
      * One of ['primary, outline-primary, secondary, danger', 'outline-danger, btn-link', btn-link-danger ]
      */
    appearance: {
      type: String,
      default: 'secondary',
      validator: function (value) {
        return Object.values(appearances).indexOf(value) !== -1
      }
    },
    /**
      * Size variations
      * One of ['default', 'small' ]
      */
    size: {
      type: String,
      default: 'default',
      validator: function (value) {
        return [
          'default',
          'small'
        ].indexOf(value) !== -1
      }
    },
    /**
     * Route object or path. If object will render <router-link>, if string
     will render <a>
     */
    to: {
      type: [Object, String],
      default: null
    },
    type: {
      type: String,
      default: 'button'
    }
  },

  computed: {
    listeners () {
      return {
        ...this.$listeners
      }
    },

    hasIcon () {
      return this.$slots.icon
    },

    hasText () {
      return this.$slots.default
    },

    buttonType () {
      return this.to ? 'router-link' : 'button'
    }
  },

  mounted () {
    this.hasIcon && this.$slots.icon[0].elm.setAttribute('viewBox', '0 0 16 16')
  }
}
</script>

<style scoped lang="scss">
@import '~@kongponents/styles/_variables.scss';

.button {
  display: inline-flex;
  align-items: center;
  padding: var(--spacing-xs, spacing(xs)) var(--spacing-sm, spacing(sm));
  font-family: var(--font-family-sans, font(sans));
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.25;
  text-decoration: none;
  color: var(--tblack-70, color(tblack-70));
  border: 1px solid transparent;
  border-radius: 3px;
  transition: all .2s ease-in-out;
  cursor: pointer;

  &:focus {
    outline: none;
  }

  &:disabled {
    opacity: 0.3;
    cursor: not-allowed;
  }

  /* Button w/ Icon */
  > svg {
    width: 1rem;
    height: 1rem;
    padding-right: var(--spacing-xs, spacing(xs));
  }

  &.icon-btn {
    width: 38px;
    height: 38px;
    padding: 0;
    justify-content: center;
    > svg {
      padding-right: 0;
    }
  }

  /* Size Variations */
  &.small {
    padding: var(--spacing-xxs, spacing(xxs)) var(--spacing-xs, spacing(xs));
    font-size: var(--type-sm, type(sm));
  }

  /* Apperance Variations */
  &.secondary {
    border-color: var(--KButtonSecondaryBorder, color(grey-88));
    background-color: var(--KButtonSecondaryBase, var(--secondary, color(grey-98)));
    &:hover {
      border-color: var(--KButtonSecondaryHoverBorder, darken(color(grey-88), 4%));
      background-color: var(--KButtonSecondaryHover, darken(color(grey-98), 4%));
    }
    &:active {
      border-color: var(--KButtonSecondaryActiveBorder, darken(color(grey-88), 8%));
      background-color: var(--KButtonSecondaryActive, darken(color(grey-98), 8%));
    }
  }
  &.primary {
    font-weight: 500;
    color: var(--twhite-1, #fff);
    background-color: var(--KButtonPrimaryBase, var(--primary, color(blue-base)));
    &:hover {
      background-color: var(--KButtonPrimaryHover, lighten(color(blue-base), 12%));
    }
    &:active {
      background-color: var(--KButtonPrimaryActive, darken(color(blue-base), 8%));
    }
  }
  &.danger {
    font-weight: 500;
    color: var(--twhite-1, #fff);
    background-color: var(--KButtonDangerBase, var(--danger, color(red-base)));
    &:hover {
      background-color: var(--KButtonDangerHover, lighten(color(red-base), 12%));
    }
    &:active {
      background-color: var(--KButtonDangerActive, darken(color(red-base), 8%));
    }
  }
  &.outline-primary {
    color: var(--KButtonLink, var(--btnLink, color(blue-link)));
    border-color: var(--KButtonOutlinePrimaryBorder, color(blue-light-01));
    background-color: var(--KButtonOutlineBackground, #fff);
    &:hover {
      background-color: var(--KButtonOutlinePrimaryHover, color(blue-lightest));
    }
    &:active {
      background-color: var(--KButtonOutlinePrimaryActive, darken(color(blue-lightest), 4%));
    }
  }
  &.outline-danger {
    color: var(--KButtonLinkDanger, var(--linkDanger, color(red-link)));
    border-color: var(--KButtonOutlineDangerBorder, color(red-light-01));
    background-color: var(--KButtonOutlineBackground, #fff);
    &:hover {
      background-color: var(--KButtonOutlineDangerHover, color(red-lightest));
    }
    &:active {
      background-color: var(--KbuttonOtlineDangerActive, darken(color(red-lightest), 4%));
    }
  }
  &.btn-link {
    color: var(--KButtonBtnLink, var(--linkPrimary, color(blue-link)));
    &:hover {
      text-decoration: underline;
    }
  }
  &.btn-link-danger {
    color: var(--KButtonLinkDanger, var(--linkDanger, color(red-link)));
    &:hover {
      text-decoration: underline;
    }
  }
}

</style>
