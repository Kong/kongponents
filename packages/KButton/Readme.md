```js
const buttonAttributes = {
  id: 'someIDName',
}
<KButton @click='alert("hey")' appearance='primary' :isRounded='true' :buttonAttributes=buttonAttributes>button</KButton>
```
