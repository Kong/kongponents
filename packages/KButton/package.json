{
  "name": "@kongponents/kbutton",
  "version": "0.1.9",
  "description": "Button component",
  "main": "KButton.vue",
  "scripts": {
    "test": "echo \"Error: no test specified\""
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "peerDependencies": {
    "@kongponents/styles": "0.0.2"
  }
}
