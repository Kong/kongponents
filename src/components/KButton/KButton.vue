<template>
  <a
    v-if="typeof to === 'string'"
    :type="type"
    :href="to"
    :class="[size, {'icon-btn': !hasText && hasIcon, 'rounded': isRounded}, appearance]"
    class="k-button"
    v-bind="$attrs"
  >

    <slot name="icon" />
    <slot />

    <KIcon
      v-if="isOpen !== undefined"
      :class="[caretClasses]"
      color="white"
      view-box="2 2 15 15"
      size="16"
      icon="chevronDown"
    />
  </a>

  <component
    :is="buttonType"
    v-else
    :type="type"
    :to="to"
    :class="[size, {'icon-btn': !hasText && hasIcon, 'rounded': isRounded}, appearance, caretClasses]"
    class="k-button"
    v-bind="$attrs"
  >
    <slot name="icon" />
    <slot />

    <KIcon
      v-if="isOpen !== undefined"
      :class="['caret', caretClasses]"
      :color="iconColor"
      view-box="2 2 15 15"
      size="16"
      icon="chevronDown"
    />
  </component>
</template>

<script lang="ts">
import { defineComponent, computed } from 'vue'
import KIcon from '@/components/KIcon/KIcon.vue'

export const appearances = {
  primary: 'primary',
  secondary: 'secondary',
  danger: 'danger',
  creation: 'creation',
  outline: 'outline',
  btnLink: 'btn-link',
}

export const sizes = {
  small: 'small',
  medium: 'medium',
  large: 'large',
}

export default defineComponent({
  name: 'KButton',
  components: { KIcon },
  inheritAttrs: false,
  props: {
    /**
      * Base styling of the button
      * One of ['primary, secondary, 'danger', 'creation', 'outline, btn-link' ]
      */
    appearance: {
      type: String,
      default: 'outline',
      validator: (value: string): boolean => {
        return Object.values(appearances).indexOf(value) !== -1
      },
    },
    /**
      * Size variations
      * One of ['default', 'small', 'medium', 'large' ]
      */
    size: {
      type: String,
      default: 'medium',
      validator: (value: string): boolean => {
        return Object.values(sizes).indexOf(value) !== -1
      },
    },
    /**
     * Route object or path. If object will render <router-link>, if string
     will render <a>
     */
    to: {
      type: [Object, String],
      default: null,
    },
    type: {
      type: String,
      default: 'button',
    },
    isOpen: {
      type: Boolean,
      default: undefined,
    },
    isRounded: {
      type: Boolean,
      default: true,
    },
  },

  setup(props, { slots }) {
    const caretClasses = computed((): string | null => {
      if (props.isOpen === undefined) return null

      return props.isOpen ? 'has-caret is-active' : 'has-caret'
    })

    const hasIcon = computed((): boolean => !!slots.icon)

    const hasText = computed((): boolean => !!slots.default)

    const buttonType = computed((): string => props.to ? 'router-link' : 'button')

    const iconColor = computed((): string => {
      if (['primary', 'danger', 'creation'].includes(props.appearance)) {
        return 'white'
      } else if (props.appearance === 'secondary') {
        return 'var(--KButtonSecondaryColor, var(--blue-600, color(blue-600)))'
      } else if (props.appearance === 'outline') {
        return 'var(--KButtonOutlineColor, var(--blue-500, color(blue-500)))'
      } else if (props.appearance === 'btn-link-danger') {
        return 'var(--KButtonLinkDanger, var(--red-500, color(red-500)))'
      }
      return ''
    })

    return {
      caretClasses,
      hasText,
      hasIcon,
      buttonType,
      iconColor,
    }
  },
})
</script>

<style lang="scss" scoped>
@import '@/styles/variables';

@mixin boxShadow($color, $whiteShadowSpred: 2px, $colorShadowSpread: 4px) {
  box-shadow: 0 0 0 $whiteShadowSpred var(--white, color(white)), 0 0 0 $colorShadowSpread $color;
}

.k-button {
  position: relative;
  display: inline-flex;
  align-items: center;
  padding: var(--KButtonPaddingY, var(--spacing-sm, spacing(sm))) var(--KButtonPaddingX, var(--spacing-lg, spacing(lg)));
  font-family: var(--font-family-sans, font(sans));
  font-size: var(--KButtonFontSize, var(--type-md, type(md)));
  font-weight: 400;
  line-height: 1.25;
  text-decoration: none;
  vertical-align: middle;
  color: var(--black-70, color(black-70));
  border: 1px solid transparent;
  border-radius: var(--KButtonRadius, 3px);
  transition: all .2s ease-in-out;
  cursor: pointer;

  &:disabled,
  &[disabled] {
    opacity: 0.3;
    cursor: not-allowed;
    &:hover {
      // ! is needed to ensure any type (a, button) keeps disabled state
      text-decoration: none !important;
    }
  }

  // remove pointer events from only <a>
  &[disabled]:not(:disabled) {
    pointer-events: none;
  }

  &:focus {
    outline: none;
  }

  /* Button w/ Icon */
  > :deep(.kong-icon) {
    padding-right: var(--spacing-xs, spacing(xs));
    box-sizing: unset;
  }

  &.icon-btn {
    height: 38px;
    justify-content: center;
    > :deep(.kong-icon) {
      padding-right: 0;
    }
  }

  /* Size Variations */
  &.small {
    padding: var(--spacing-xs, spacing(xs)) var(--spacing-md, spacing(md));
    font-size: var(--KButtonFontSize, 13px);
    line-height: 13px;
  }

  &.medium {
    padding: var(--spacing-sm, spacing(sm)) var(--spacing-lg, spacing(lg)) ;
    font-size: var(--KButtonFontSize, var(--type-md, type(md)));
    line-height: var(--type-md, type(md));
  }

  &.large {
    padding: var(--spacing-md, spacing(md)) var(--spacing-xl, spacing(xl)) ;
    font-size: var(--KButtonFontSize, var(--type-md, type(md)));
    line-height: var(--type-md, type(md));
  }

  /* class to add for dropdown caret */

  :deep(.caret) {
    margin-left: 15px;
    padding: 0;
    display: inline-block;
    transition: 250ms ease;
  }

  &.is-active :deep(.caret) {
    transform: rotate(-180deg);
    transition: 250ms ease;
  }

  /* Apperance Variations */
  &.secondary {
    color: var(--KButtonSecondaryColor, var(--blue-600, color(blue-600)));
    background-color: var(--KButtonSecondaryBase, var(--blue-200, color(blue-200)));
    &:hover:not(:disabled) {
      background-color: var(--KButtonSecondaryHover, var(--blue-300));
    }
    &:active {
      background-color: var(--KButtonSecondaryActive, var(--blue-300, color(blue-300)));
    }
    &:focus {
      @include boxShadow(var(--KButtonSecondaryBase, var(--blue-300, color(blue-300))));
    }
  }

  &.primary {
    color: var(--white, #fff);
    background-color: var(--KButtonPrimaryBase, var(--blue-500, color(blue-500)));
    &:hover:not(:disabled) {
      background-color: var(--KButtonPrimaryHover, var(--blue-600));
    }
    &:active {
      background-color: var(--KButtonPrimaryActive, var(--blue-600, color(blue-600)));
    }
    &:focus {
      @include boxShadow(var(--KButtonPrimaryBase, var(--blue-500, color(blue-500))));
    }
  }
  &.danger {
    color: var(--white, #fff);
    background-color: var(--KButtonDangerBase, var(--red-500, color(red-500)));
    &:hover:not(:disabled) {
      $hover: rgba(color(red-700), .85);
      background-color: var(--KButtonDangerHover, $hover);
    }
    &:active {
      background-color: var(--KButtonDangerActive, var(--red-700, color(red-700)));
    }
    &:focus {
      @include boxShadow(var(--KButtonDangerBase, var(--red-700, color(red-700))));
    }
  }

  &.creation {
    color: var(--white, #fff);
    background-color: var(--KButtonCreationBase, var(--green-400, color(green-400)));
    &:hover:not(:disabled) {
      $hover: rgba(color(green-500), .85);
      background-color: var(--KButtonCreationHover, $hover);
    }
    &:active {
      background-color: var(--KButtonCreationActive, var(--green-400, color(green-400)));
    }
    &:focus {
      @include boxShadow(var(--KButtonCreationBase, var(--green-500, color(green-500))));
    }
  }

  &.outline {
    color: var(--KButtonOutlineColor, var(--blue-500, color(blue-500)));
    border-color: var(--KButtonOutlineBorder, rgba(color(blue-500), .4));
    background-color: var(--white, color(white));
    outline-style: inherit;
    &:hover:not(:disabled) {
      border-color: var(--KButtonOutlineHoverBorder, rgba(color(blue-500), 1));
    }
    &:active {
      border-color: var(--KButtonOutlineActiveBorder, rgba(color(blue-500), 1));
      background-color: var(--KButtonOutlineActive, var(--blue-100, color(blue-100)));
    }
    &:focus {
      @include boxShadow(var(--KButtonOutlineBorder, var(--blue-500, color(blue-500))));
    }
  }
  &.btn-link {
    color: var(--KButtonBtnLink, var(--blue-500, color(blue-500)));
    background-color: transparent;
    &:hover:not(:disabled) {
      text-decoration: underline;
    }
    &:focus {
      @include boxShadow(var(--blue-500, color(blue-500)), 0, 2px);
    }
  }

  &.btn-link-danger {
    color: var(--KButtonLinkDanger, var(--red-500, color(red-500)));
    background-color: transparent;
    &:hover:not(:disabled) {
      text-decoration: underline;
    }
    &:focus {
      @include boxShadow(var(--red-700, color(red-700)), 0, 2px);
    }
  }

  &.rounded {
    border-radius: 100px;
  }
}
</style>

<style lang="scss">
.k-button.btn-link.has-caret .caret.has-caret path {
  stroke: var(--KButtonBtnLink, var(--blue-500, color(blue-500)));
}
</style>
