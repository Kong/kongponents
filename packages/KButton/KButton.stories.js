import Vue from 'vue'
import { storiesOf } from '../../storybook/node_modules/@storybook/vue'
import { action } from '../../storybook/node_modules/@storybook/addon-actions'
import { withInfo } from '../../storybook/node_modules/storybook-addon-vue-info'

import { default as KButton, appearances } from './KButton.vue'
import KIcon from '../KIcon/KIcon.vue'

storiesOf('KButton', module)
  .add('Default', () => ({
    components: { KButton },
    template: `<KButton>button</KButton>`
  }), {info: {}})
  .add('Anchor Link', () => ({
    components: { KButton },
    template: `<div>
      <KButton appearance="btn-link" :to="{path: '/'}">Router Link!</KButton>
      <KButton appearance="btn-link" to="http://google.com">Anchor Link!</KButton>
    </div>`
  }), {info: {}})
  .add('Appearance', () => ({
    components: { KButton },
    template: `
      <div>
      ${Object.values(appearances).map(type => {
      return `<KButton appearance="${type}" style="margin: 0 5px">${type}</KButton>`
    }).join('')}
      </div>`
  }), {info: {}})
  .add('Disabled', () => ({
    components: { KButton },
    template: `<div>
      ${Object.values(appearances).map(type => {
      return `<KButton appearance="${type}" disabled style="margin: 0 5px">${type}</KButton>`
    }).join('')}
      </div>`
  }), {info: {}})
  .add('Sizes', () => ({
    components: { KButton },
    template: `<div>
      <KButton appearance="secondary" size="small">small</KButton>
    </div>`
  }), { info: {} })
  .add('Icons', () => ({
    components: { KButton, KIcon },
    template: `<div>
      <KButton appearance="primary">
        <KIcon slot="icon" icon="spinner" />
        With Text
      </KButton>
      <KButton appearance="primary">
        <KIcon slot="icon" icon="spinner"/>
      </KButton>
      <KButton appearance="primary">
        With Text
      </KButton>
    </div>`
  }), {info: {}})
  .add('Actions', () => ({
    components: { KButton },
    template: `
    <KButton
      @click="onClick"
      @focus="onFocus"
      @blur="onBlur"
      @dblclick="onDblClick"
      appearance='primary'>
      button
    </KButton>`,
    methods: {
      onClick (e) {
        action('onClick')(e)
      },
      onFocus (e) {
        action('onFocus')(e)
      },
      onBlur (e) {
        action('onBlur')(e)
      },
      onDblClick (e) {
        action('onDblClick')(e)
      }
    }
  }), {info: {}})
