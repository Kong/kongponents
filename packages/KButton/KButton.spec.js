import { mount, createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import { default as KButton, appearances } from '@/KButton/KButton'

const rendersCorrectAppearance = (variant) => {
  it(`renders kbutton with the ${variant} appearance`, () => {
    const wrapper = mount(KButton, {
      propsData: {
        'appearance': variant
      },
      slots: {
        default: variant
      }
    })

    expect(wrapper.find('.button').classes()).toContain(`${variant}`)
    expect(wrapper.html()).toMatchSnapshot()
  })
}

describe('KButton', () => {
  Object.values(appearances).map(a => rendersCorrectAppearance(a))

  it('sets small class when size passed', () => {
    const wrapper = mount(KButton, {
      propsData: {
        'size': 'small'
      },
      slots: {
        default: 'Small Button'
      }
    })

    expect(wrapper.find('.button').classes()).toContain('small')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders a router link with KButton styles', () => {
    const localVue = createLocalVue()
    const router = new VueRouter({
      mode: 'history',
      routes: [{
        path: 'services',
        name: 'services',
        redirect: 'services',
        meta: { breadcrumb: 'Services' }
      }]
    })

    localVue.use(VueRouter)
    const wrapper = mount(KButton, {
      localVue,
      router,
      propsData: {
        'to': { name: 'services', params: { workspace: 'test' } },
        appearance: 'primary'
      },
      slots: {
        default: `I'm a router link`
      }
    })

    expect(wrapper.find('a').classes()).toContain('button', 'primary')
    expect(wrapper.find('a').attributes('href')).toBe('/services')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders a native link with KButton styles', () => {
    const wrapper = mount(KButton, {
      propsData: {
        'to': 'https://google.com',
        appearance: 'secondary'
      },
      slots: {
        default: `I'm a native link`
      }
    })

    expect(wrapper.find('a').classes()).toContain('button', 'secondary')
    expect(wrapper.find('a').attributes('href')).toBe('https://google.com')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
