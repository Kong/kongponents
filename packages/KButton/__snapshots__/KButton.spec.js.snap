// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KButton renders a native link with KButton styles 1`] = `<a type="button" href="https://google.com" class="button secondary">I'm a native link</a>`;

exports[`KButton renders a router link with KButton styles 1`] = `<a href="/services" class="button primary" type="button">I'm a router link</a>`;

exports[`KButton renders kbutton with the btn-link appearance 1`] = `<button type="button" class="button btn-link">btn-link</button>`;

exports[`KButton renders kbutton with the btn-link-danger appearance 1`] = `<button type="button" class="button btn-link-danger">btn-link-danger</button>`;

exports[`KButton renders kbutton with the danger appearance 1`] = `<button type="button" class="button danger">danger</button>`;

exports[`KButton renders kbutton with the outline-danger appearance 1`] = `<button type="button" class="button outline-danger">outline-danger</button>`;

exports[`KButton renders kbutton with the outline-primary appearance 1`] = `<button type="button" class="button outline-primary">outline-primary</button>`;

exports[`KButton renders kbutton with the primary appearance 1`] = `<button type="button" class="button primary">primary</button>`;

exports[`KButton renders kbutton with the secondary appearance 1`] = `<button type="button" class="button secondary">secondary</button>`;

exports[`KButton sets small class when size passed 1`] = `<button type="button" class="button small secondary">Small Button</button>`;
